browser/lib/util/crypto.js

```diff
@@ -98,84 +115,102 @@ var Crypto = (function() {
+			if(typeof(params) === 'function') {
+				Crypto.generateRandomKey(function(key) {
+					params(null, Crypto.getDefaultParams({key: key}));
+				})
+			} else {
+				callback = arguments[1];
+				callback(null, Crypto.getDefaultParams({key: params}));
+			}
 			return;
 		}
-		if (typeof(key) === 'string')
-			key = CryptoJS.enc.Hex.parse(key);
-		else
-			key = BufferUtils.toWordArray(key);   // Expect key to be an Array, ArrayBuffer, or WordArray at this point
-
-		var params = new CipherParams();
-		params.algorithm = DEFAULT_ALGORITHM;
-		params.key = key;
-		params.keyLength = key.words.length * (4 * 8);
-		params.mode = DEFAULT_MODE;
-		generateRandom(DEFAULT_BLOCKLENGTH, function(err, buf) {
-			params.iv = buf;
-			callback(null, params);
-		});
+
+		if(!params.key) {
+			throw new Error('Crypto.getDefaultParams: a key is required');
+		}
+
+		if (typeof(params.key) === 'string') {
+			key = CryptoJS.enc.Base64.parse(normaliseBase64(params.key));
+		} else {
+			key = BufferUtils.toWordArray(params.key); // Expect key to be an Array, ArrayBuffer, or WordArray at this point
+		}
+
+		var cipherParams = new CipherParams();
+		cipherParams.key = key;
+		cipherParams.algorithm = params.algorithm || DEFAULT_ALGORITHM;
+		cipherParams.keyLength = key.words.length * (4 * 8);
+		cipherParams.mode = params.mode || DEFAULT_MODE;
+		validateCipherParams(cipherParams);
+		return cipherParams;
 	};
 
 	/**
-	 * Internal; get a ChannelCipher instance based on the given ChannelOptions
-	 * @param channelOpts a ChannelOptions instance
-	 * @param callback (err, cipher)
+	 * Generate a random encryption key from the supplied keylength (or the
+	 * default keyLength if none supplied) as a CryptoJS WordArray
+	 * @param keyLength (optional) the required keyLength in bits
+	 * @param callback (err, key)
 	 */
-	Crypto.getCipher = function(channelOpts, callback) {
-		var params = channelOpts && channelOpts.cipherParams;
-		if(params) {
-			callback(null, new CBCCipher(params));
-			return;
+	Crypto.generateRandomKey = function(keyLength, callback) {
+		if(arguments.length == 1 && typeof(keyLength) == 'function') {
+			callback = keyLength;
+			keyLength = undefined;
+		}
+		generateRandom((keyLength || DEFAULT_KEYLENGTH) / 8, callback);
+	};
+
+	/**
+	 * Internal; get a ChannelCipher instance based on the given cipherParams
+	 * @param params either a CipherParams instance or some subset of its
+	 * fields that includes a key
+	 * @param callback (err, cipherParams, channelCipher)
+	 */
+	Crypto.getCipher = function(params, callback) {
+		var cipherParams = (params instanceof CipherParams) ?
+		                   params :
+		                   Crypto.getDefaultParams(params);
+
+		if(params.iv) {
+			callback(null, cipherParams, new CBCCipher(cipherParams, params.iv));
+		} else {
+			generateRandom(DEFAULT_BLOCKLENGTH, function(err, iv) {
+				callback(null, cipherParams, new CBCCipher(cipherParams, iv));
```

Comment:
@SimonWoolf I've not read this thoroughly, but this looks like the Cipher has an embedded `iv` and thus uses the same `iv` for every single request.  This is not the intention, the `iv` should be generated uniquely for every message encrypted, unless an `iv` is explicitly provided for testing purposes.